frame/nft-fractionalization/src/lib.rs

```diff
@@ -0,0 +1,316 @@
+			Self::do_set_metadata(asset_id, &who, &pallet_account, &nft_collection_id, &nft_id)?;
+
+			NftToAsset::<T>::insert(
+				(nft_collection_id, nft_id),
+				Details { asset: asset_id, fractions },
+			);
+
+			Self::deposit_event(Event::NftFractionalized {
+				nft_collection: nft_collection_id,
+				nft: nft_id,
+				fractions,
+				asset: asset_id,
+				beneficiary,
+			});
+
+			Ok(())
+		}
+
+		/// Burn the whole amount of the asset and return back the locked NFT.
+		#[pallet::call_index(1)]
+		#[pallet::weight(T::WeightInfo::unify())]
+		pub fn unify(
+			origin: OriginFor<T>,
+			nft_collection_id: T::NftCollectionId,
+			nft_id: T::NftId,
+			asset_id: AssetIdOf<T>,
+			beneficiary: AccountIdLookupOf<T>,
+		) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+			let beneficiary = T::Lookup::lookup(beneficiary)?;
+
+			NftToAsset::<T>::try_mutate_exists((nft_collection_id, nft_id), |maybe_details| {
+				let details = maybe_details.take().ok_or(Error::<T>::DataNotFound)?;
+				ensure!(details.asset == asset_id, Error::<T>::DataNotFound);
+
+				Self::do_burn_asset(asset_id, &who, details.fractions)?;
+				Self::do_unlock_nft(nft_collection_id, nft_id, &beneficiary)?;
+
+				Self::deposit_event(Event::NftUnified {
+					nft_collection: nft_collection_id,
+					nft: nft_id,
+					asset: asset_id,
+					beneficiary,
+				});
+
+				Ok(())
+			})
+		}
+	}
+
+	impl<T: Config> Pallet<T> {
+		/// The account ID of the pallet.
+		///
+		/// This actually does computation. If you need to keep using it, then make sure you cache
+		/// the value and only call this once.
+		fn get_pallet_account() -> T::AccountId {
+			T::PalletId::get().into_account_truncating()
+		}
+
+		/// Transfer the NFT from the account holding that NFT to the pallet's account.
+		fn do_lock_nft(nft_collection_id: T::NftCollectionId, nft_id: T::NftId) -> DispatchResult {
+			T::NftLocker::lock(&nft_collection_id, &nft_id)
+		}
+
+		/// Transfer the NFT to the account returning the tokens.
+		fn do_unlock_nft(
+			nft_collection_id: T::NftCollectionId,
+			nft_id: T::NftId,
+			account: &T::AccountId,
+		) -> DispatchResult {
+			T::NftLocker::unlock(&nft_collection_id, &nft_id)?;
+			T::Nfts::transfer(&nft_collection_id, &nft_id, account)
+		}
+
+		/// Create the new asset.
+		fn do_create_asset(asset_id: AssetIdOf<T>, admin: T::AccountId) -> DispatchResult {
+			T::Assets::create(asset_id, admin, false, One::one())
+		}
+
+		/// Mint the `amount` of tokens with `asset_id` into the beneficiary's account.
+		fn do_mint_asset(
+			asset_id: AssetIdOf<T>,
+			beneficiary: &T::AccountId,
+			amount: AssetBalanceOf<T>,
+		) -> DispatchResult {
+			T::Assets::mint_into(asset_id, beneficiary, amount)
+		}
+
+		/// Burn tokens from the account.
+		fn do_burn_asset(
+			asset_id: AssetIdOf<T>,
+			account: &T::AccountId,
+			amount: AssetBalanceOf<T>,
+		) -> DispatchResult {
+			T::Assets::burn_from(asset_id, account, amount)?;
+			T::Assets::start_destroy(asset_id, None)
```

Comment:
If we start the `start_destroy` process, than why we don't do also the `destroy_approvals` and `finish_destroy`?

Reply:
Because we don't know the number of approvals the asset has :(
For the `destroy_accounts` it's easier, as we can assume there is only one account left that holds the asset (pallet's)